### PR TITLE
[bug][algorithm] remove incorrect torch.no_grad() for kl in loss (use_kl_loss=True)

### LINKED ISSUE #1340: Gradient disabled for KL loss computation
# Description

[`compute_approx_kl`](https://github.com/NovaSky-AI/SkyRL/blob/be602d8ca3ef7f6f71cdab26acb0889e2b64c81a/skyrl/backends/skyrl_train/utils/ppo_utils.py#L90) is decorated with `torch.no_grad`, and it's used to compute the KL loss in [`worker.py`](https://github.com/NovaSky-AI/SkyRL/blob/be602d8ca3ef7f6f71cdab26acb0889e2b64c81a/skyrl/backends/skyrl_train/workers/worker.py#L886). Since there's no grad, `kl_loss` is simply a scalar that doesn't affect gradient computation when added to the loss. While using `torch.no_grad` makes sense when adding the KL to the reward, it seems incorrect when using it as a loss term through which gradients should flow.

As a sanity check, the [`kl_penalty_forward`](https://github.com/verl-project/verl/blob/322c8208e4694a872d2d2c431402e5344fbec56a/verl/trainer/ppo/core_algos.py#L2152) function in verl is not decorated with `torch.no_grad`, neither at the function definition nor where it's used in [`kl_penalty`](https://github.com/verl-project/verl/blob/322c8208e4694a872d2d2c431402e5344fbec56a/verl/trainer/ppo/core_algos.py#L2138) and [`ppo_loss`](https://github.com/verl-project/verl/blob/322c8208e4694a872d2d2c431402e5344fbec56a/verl/workers/utils/losses.py#L97).

# Proposed Solution

1. Remove the `torch.no_grad` decorator from `compute_approx_kl`.
2. Call `compute_approx_kl` in a `with torch.no_grad()` block in `trainer.py` in [`apply_reward_kl_penalty`](https://github.com/NovaSky-AI/SkyRL/blob/be602d8ca3ef7f6f71cdab26acb0889e2b64c81a/skyrl/train/trainer.py#L998).

### PR DESCRIPTION
**Pull request by erictang000**
Fixes #1340

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1353" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by erictang000** [comment]
GSM8K with and without fix (looks pretty similar, but probably just b/c it's gsm8k):
image

**Comment by SumanthRH** [comment]
@erictang000 should be able to see the diff with a large kl penalty. Could also just print the kl loss  tensor in the worker before and after the fix (after should have requires grad)

**Comment by erictang000** [comment]
here after setting `kl_loss_coef=1.0`

image
